ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -91,69 +159,396 @@ def loop(self):
-                          "from {} to {}".format(old_update_rate,
-                                                 config['dyn_update_rate']))
-            self.update_rate = config['dyn_update_rate']
-            # need to switch the delay
-            self.rate = rospy.Rate(self.update_rate)
-        # end if
+    # Receive a msg from /traffic_waypoint about the next stop line
+    def traffic_cb(self, traffic_msg):
+        if traffic_msg.data != self.next_traffic_light_wp:
+            self.next_traffic_light_wp = traffic_msg.data
+            rospy.loginfo("new /traffic_waypoint message received at wp: %d."
+                          "while car is at wp %d", self.next_traffic_light_wp,
+                          self.final_waypoints_start_ptr)
+        else:
+            # just for debug to see what we're getting
+            rospy.logdebug("same /traffic_waypoint message received.")
+        # end if else
 
-        if old_default_velocity != config['dyn_default_velocity']:
-            rospy.loginfo("waypoint_updater:dyn_vars_cb Adjusting default_"
-                          "velocity from {} to {}"
-                          .format(old_default_velocity,
-                                  config['dyn_default_velocity']))
+    def send_waypoints(self):
+        # generates the list of LOOKAHEAD_WPS waypoints based on car location
+        # for now assume waypoints form a loop
+        # waypoints loaded into a deque with deepcopy
+        # buffer of rear_buffer points to use for location search
+        # final_waypoints_start_ptr refers to waypoint_ptr which will be start
+        # of published final_waypoints list
+        # last_waypoint was the initial waypoint in previous iteration
 
-            if config['dyn_default_velocity'] > self.max_velocity:
-                rospy.logwarn("waypoint_updater:dyn_vars_cb default_velocity "
-                              "limited to max_velocity {}"
-                              .format(self.max_velocity))
-                self.default_velocity = self.max_velocity
+        start_wps_ptr = self.closest_waypoint()
+        dequelen = len(self.f_wpdeq)
+        if self.final_waypoints_start_ptr < 0:
+            # condition on startup
+            self.final_waypoints_start_ptr = start_wps_ptr
+            self.last_waypoint = self.final_waypoints_start_ptr
+            end_waypoint_in_deque = self.last_waypoint - 1
+            # load buffer of points behind into deque
+            for w_ptr in range(end_waypoint_in_deque - self.rear_buffer,
+                               end_waypoint_in_deque+1):
+                wpd = deq_waypoint(w_ptr, copy.deepcopy(self.waypoints[w_ptr]))
+                self.f_wpdeq.append(wpd)
+        else:
+            self.last_waypoint = self.final_waypoints_start_ptr
+            end_waypoint_in_deque = self.f_wpdeq[dequelen - 1].\
+                ptr_id
+            self.final_waypoints_start_ptr = start_wps_ptr
```

Comment:
It also looks like self.final_waypoints_start_ptr is always set = start_wps_ptr = self.closest_waypoint().  Can we just directly set `self.final_waypoints_start_ptr = self.closest_waypoint()` at the top in line 190 (no need for temp variable start_wps_ptr)?

Reply:
Hi
I'm just going to post new simpler proof of concept test code that I'm working on which should work and solve most problems.  Don't go line by line right now on this - It is a big spaghetti mess.